src/radiation/moments_source.cpp

```diff
@@ -0,0 +1,801 @@
+  auto eos = eos_pkg->Param<singularity::EOS>("d.EOS");
+  namespace cr = radmoment_cons;
+  namespace pr = radmoment_prim;
+  namespace ir = radmoment_internal;
+  namespace c = fluid_cons;
+  namespace p = fluid_prim;
+  std::vector<std::string> vars{c::energy, c::momentum, c::ye,       cr::E,
+                                cr::F,     c::bfield,   p::density,  p::temperature,
+                                p::energy, p::ye,       p::velocity, p::bfield,
+                                pr::J,     pr::H,       ir::kappaJ,  ir::kappaH,
+                                ir::JBB,   ir::tilPi,   ir::srcfail};
+
+  PackIndexMap imap;
+  auto v = rc->PackVariables(vars, imap);
+  auto idx_E = imap.GetFlatIdx(cr::E);
+  auto idx_F = imap.GetFlatIdx(cr::F);
+  auto idx_J = imap.GetFlatIdx(pr::J);
+  auto idx_H = imap.GetFlatIdx(pr::H);
+
+  auto idx_kappaJ = imap.GetFlatIdx(ir::kappaJ);
+  auto idx_kappaH = imap.GetFlatIdx(ir::kappaH);
+  auto idx_JBB = imap.GetFlatIdx(ir::JBB);
+  auto idx_tilPi = imap.GetFlatIdx(ir::tilPi, false);
+  auto ifail = imap[ir::srcfail].first;
+  auto pv = imap.GetFlatIdx(p::velocity);
+
+  int prho = imap[p::density].first;
+  int peng = imap[p::energy].first;
+  int pT = imap[p::temperature].first;
+  int pYe = imap[p::ye].first;
+  int pb_lo = imap[p::bfield].first;
+  int cb_lo = imap[c::bfield].first;
+
+  int ceng = imap[c::energy].first;
+  int cmom_lo = imap[c::momentum].first;
+  int cmom_hi = imap[c::momentum].second;
+  int cye = imap[c::ye].first;
+
+  IndexRange ib = rc->GetBoundsI(IndexDomain::interior);
+  IndexRange jb = rc->GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = rc->GetBoundsK(IndexDomain::interior);
+
+  // Get the background geometry
+  auto geom = Geometry::GetCoordinateSystem(rc);
+
+  int nblock = v.GetDim(5);
+
+  auto species = rad->Param<std::vector<RadiationType>>("species");
+  auto num_species = rad->Param<int>("num_species");
+  RadiationType species_d[3] = {};
+  for (int s = 0; s < num_species; s++) {
+    species_d[s] = species[s];
+  }
+
+  auto bounds = fix_pkg->Param<fixup::Bounds>("bounds");
+
+  auto coords = pmb->coords;
+
+  const auto &d_opacity = opac->Param<Opacity>("d.opacity");
+  const auto &d_mean_opacity = opac->Param<MeanOpacity>("d.mean_opacity");
+  // Mainly for testing purposes, probably should be able to do this with the opacity code
+  // itself
+  const auto scattering_fraction = rad->Param<Real>("scattering_fraction");
+
+  const auto src_solver = rad->Param<SourceSolver>("src_solver");
+  const auto src_use_oned_backup = rad->Param<bool>("src_use_oned_backup");
+  const auto src_rootfind_eps = rad->Param<Real>("src_rootfind_eps");
+  const auto src_rootfind_tol = rad->Param<Real>("src_rootfind_tol");
+  const auto src_rootfind_maxiter = rad->Param<int>("src_rootfind_maxiter");
+  const auto oned_fixup_strategy = rad->Param<OneDFixupStrategy>("oned_fixup_strategy");
+
+  parthenon::par_for(
```

Comment:
This kernel is an absolute beast, and difficult to parse. Anyway to break it up or at least have some device functions to call?

Reply:
I agree, I held off on doing that though because both 1D and 4D solvers require a lot of data that would have to become function arguments. Though maybe I could make a `struct` they share. I could probably reduce code sharing between the different 1D solvers (`regular` and `ignore_dJ`) a bit at least.

Reply:
OK thinking more about this I think the right thing to do is a semi-involved refactor (#131) which I would like to defer for a little while. But I'll definitely get to this by the time I add the `twod` solver. That OK?

Reply:
:+1: works for me